services/github/github-commit-status.service.js

```diff
@@ -35,71 +27,56 @@ module.exports = class GithubCommitStatus extends LegacyService {
-          color: 'brightgreen',
-        },
+        staticPreview: this.render({
+          isInBranch: true,
+          branch: 'master',
+        }),
         keywords: ['branch'],
         documentation,
       },
     ]
   }
 
-  static registerLegacyRouteHandler({ camp, cache, githubApiProvider }) {
-    camp.route(
-      /^\/github\/commit-status\/([^/]+)\/([^/]+)\/([^/]+)\/([^/]+)\.(svg|png|gif|jpg|json)$/,
-      cache((data, match, sendBadge, request) => {
-        const [, user, repo, branch, commit, format] = match
-        const apiUrl = `/repos/${user}/${repo}/compare/${branch}...${commit}`
-        const badgeData = getBadgeData('commit status', data)
-        githubApiProvider.request(request, apiUrl, {}, (err, res, buffer) => {
-          if (
-            githubCheckErrorResponse(
-              badgeData,
-              err,
-              res,
-              'commit or branch not found'
-            )
-          ) {
-            if (res && res.statusCode === 404) {
-              try {
-                if (
-                  JSON.parse(buffer).message.startsWith(
-                    'No common ancestor between'
-                  )
-                ) {
-                  badgeData.text[1] = 'no common ancestor'
-                  badgeData.colorscheme = 'lightgrey'
-                }
-              } catch (e) {
-                badgeData.text[1] = 'invalid'
-                badgeData.colorscheme = 'lightgrey'
-              }
-            }
-            sendBadge(format, badgeData)
-            return
-          }
-          try {
-            const parsedData = JSON.parse(buffer)
-            const isInBranch =
-              parsedData.status === 'identical' ||
-              parsedData.status === 'behind'
-            if (isInBranch) {
-              badgeData.text[1] = `in ${branch}`
-              badgeData.colorscheme = 'brightgreen'
-            } else {
-              // status: ahead or diverged
-              badgeData.text[1] = `not in ${branch}`
-              badgeData.colorscheme = 'yellow'
-            }
-            sendBadge(format, badgeData)
-          } catch (e) {
-            badgeData.text[1] = 'invalid'
-            sendBadge(format, badgeData)
-          }
-        })
-      })
-    )
+  static get defaultBadgeData() {
+    return {
+      label: 'commit status',
+    }
+  }
+
+  static render({ isInBranch, branch }) {
+    if (isInBranch) {
+      return {
+        message: `in ${branch}`,
+        color: 'brightgreen',
+      }
+    } else {
+      // status: ahead or diverged
+      return {
+        message: `not in ${branch}`,
+        color: 'yellow',
+      }
+    }
+  }
+
+  async handle({ user, repo, branch, commit }) {
+    let status
+    try {
+      ;({ status } = await this._requestJson({
+        url: `/repos/${user}/${repo}/compare/${branch}...${commit}`,
+        errorMessages: errorMessagesFor('commit or branch not found'),
+        schema: Joi.object().required(),
```

Comment:
If the schema is vague here so that we pass the success response and error response, could we do something like `Joi.alternatives(errorSchema, validSchema)`?

Reply:
Oops, I meant to fill that in.

Declaring the error schema doesn't do any good because _requestJson won't even parse the response on a 404.